compiler/src/iree/compiler/Codegen/Common/GPU/GPUMaterializeEncoding.cpp

```diff
@@ -319,6 +328,146 @@ struct GPUSetEncodingOpLoweringConversion
+    AffineExpr mExpr = rewriter.getAffineDimExpr(0);
+    AffineExpr nExpr = rewriter.getAffineDimExpr(1);
+    AffineExpr kExpr = rewriter.getAffineDimExpr(2);
+
+    // The outer dims are all in row-major fasion after relayout.
+    auto lhsMap = AffineMap::get(3, 0, {mExpr, kExpr}, ctx);
+    auto rhsMap = AffineMap::get(3, 0, {nExpr, kExpr}, ctx);
+    auto accMap = AffineMap::get(3, 0, {mExpr, nExpr}, ctx);
+
+    SmallVector<utils::IteratorType> iteratorTypes =
+        linalgOp.getIteratorTypesArray();
+
+    // TODO(hanchung): Support batch gemms.
+    // The current multi_mma op is not able to represent the data-tiled mma
+    // layouts. There are internal details about layout after we data-tile a
+    // matrix. The matrix is in row-major fashion for outer dimensions. The
+    // inner dimensions were in row-major, and then we swizzle them to be
+    // column-major. The number of inner dimensions is greater than 2 after tile
+    // swizzling. Here we use the "row-major" layout for inner dimensions. The
+    // shape is exacatly innerTile[0]xinnerTile[1] where innerTile can be
+    // derived from the encoding info. The chunk of the inner tiles are in
+    // column-major fashion, but it does not really matter. Because they are
+    // opaque. I don't find a way to represent it, so I added a
+    // `__is_data_tiled__` magic attribute to the op. The expectation is that
+    // there is a pass to resolve the `__is_data_tiled__` attribute before
+    // additional lowering. It should unroll the op and resolve the layout; each
+    // multi_mma op maps to a single intrisic at subgroup level. Then we do the
+    // lowering that we've been doing in TileAndFuse pipeline.
+    //
+    // E.g., say that the intrinsic is MFMA_f32_16x16x4_f32, and the unrolling
+    // factors for [M, N, K] are [2, 4, 1], then we have:
+    //
+    //   iree_gpu.multi_mma %lhs, %rhs, %acc {
+    //     kind = #iree_gpu.mma_layout<MFMA_F32_16x16x4_F32>,
+    //     lhs_permutation = array<i64: 0, 1>,
+    //     rhs_permutation = array<i64: 0, 1>,
+    //     __is_data_tiled__
+    //   } : tensor<?x?x32x4xf32>, tensor<?x?x16x4xf32>
+    //     into tensor<?x?x32x64xf32>
+    //
+    // The data-tiled chunk represents in row-major favor, so the permutation is
+    // identiy; the `__is_data_tiled__` attribute is attached.
+    //
+    // Then we resolve the `__is_data_tiled__` layout, which do the unrolling.
+    // It becomes a sequence of
+    //
+    //   iree_gpu.multi_mma %lhs_i, %rhs_i, %acc_i {
+    //     kind = #iree_gpu.mma_layout<MFMA_F32_16x16x4_F32>,
+    //     lhs_permutation = array<i64: 1, 0>,
+    //     rhs_permutation = array<i64: 1, 0>,
+    //   } : tensor<?x?x4x16xf32>, tensor<?x?x4x16xf32>
+    //     into tensor<?x?x16x16xf32>
+    //
+    // This is what we usually have in TileAndFuse pipeline. IMO, this step
+    // happens after we distribute the op to subgroups/threads; before we
+    // distribute them to lanes.
```

Comment:
@Max191 @qedawkins I took a stab at writing down one of potential path for data-tiled multi_mma op. The idea is documented in the comment, and I will try to prototype that. Let me know if you see something weird and we can discuss it next week.

(cc @lialan @pashu123 @bjacob @MaheshRavishankar )

Reply:
```
    // Here we use the "row-major" layout for inner dimensions. The
    // shape is exactly innerTile[0]xinnerTile[1] where innerTile can be
    // derived from the encoding info. The chunk of the inner tiles are in
    // column-major fashion, but it does not really matter.
``` 
@hanhanW Could you explain this part?

Reply:
The shapes here don't match what I had in my head:
```
    // E.g., say that the intrinsic is MFMA_f32_16x16x4_f32, and the unrolling
    // factors for [M, N, K] are [2, 4, 1], then we have:
    //
    //   iree_gpu.multi_mma %lhs, %rhs, %acc {
    //     kind = #iree_gpu.mma_layout<MFMA_F32_16x16x4_F32>,
    //     lhs_permutation = array<i64: 0, 1>,
    //     rhs_permutation = array<i64: 0, 1>,
    //     __is_data_tiled__
    //   } : tensor<?x?x32x4xf32>, tensor<?x?x16x4xf32>
    //     into tensor<?x?x32x64xf32>
```
My understanding was that the interleaved dimension will always be collapsed into the innermost dimension. So, in this example, with `[2, 4, 1]` unrolling, the shapes would look like:
```
    // E.g., say that the intrinsic is MFMA_f32_16x16x4_f32, and the unrolling
    // factors for [M, N, K] are [2, 4, 1], then we have:
    //
    //   iree_gpu.multi_mma %lhs, %rhs, %acc {
    //     kind = #iree_gpu.mma_layout<MFMA_F32_16x16x4_F32>,
    //     lhs_permutation = array<i64: 1, 0>,
    //     rhs_permutation = array<i64: 1, 0>,
    //     __is_data_tiled__
    //   } : tensor<?x?x4x32xf32>, tensor<?x?x16x16xf32>
    //     into tensor<?x?x32x64xf32>
```
Which becomes a sequence of 8 single tiles:
```
    // E.g., say that the intrinsic is MFMA_f32_16x16x4_f32, and the unrolling
    // factors for [M, N, K] are [2, 4, 1], then we have:
    //
    //   iree_gpu.multi_mma %lhs, %rhs, %acc {
    //     kind = #iree_gpu.mma_layout<MFMA_F32_16x16x4_F32>,
    //     lhs_permutation = array<i64: 1, 0>,
    //     rhs_permutation = array<i64: 1, 0>,
    //     __is_data_tiled__
    //   } : tensor<?x?x4x16xf32>, tensor<?x?x16x4xf32>
    //     into tensor<?x?x16x16xf32>
```
As I think about it more, I think we actually only need the `__is_data_tiled__` to be able to infer the interleaved unrolling factors. Unrolling of M and N can be determined from the output shape, and the interleaved unrolling factor can be determined by the size of the inner dimensions on the LHS and RHS.

Reply:
I talked to Max offline and we brainstorm what is missing and what needs to be done. I was over anchoring on the compatibility of CPU and abusing the opaque types in multi_mma op, which makes the inner tiles shape tricky. We can just add "resolveEncodingShape" to the type converter or whatever, and produce the shape in those binding ops. Then we don't need [additional reshapes](https://github.com/iree-org/iree/blob/cd78e91a2953855840bc09dbed8e45cf05846aaf/compiler/src/iree/compiler/Codegen/Common/GPU/test/gpu_materialize_encoding.mlir#L42-L45) to make flow/hal binding ops happy.

The issue is that how do we represent the materialized contraction ops. There is no existing op that can express the op. The closest one is `multi_mma` op, but the definition does not support n-D inner tiles. Basically the `*_permutation` attributes are broken in the case. The op we're looking for is something similar to `multi_mma` without those permutation attributes. We think that we can still attach `__is_data_tiled__` to the op and say that it has different semantics on inner tiles. The op is very useful because it already implements TilingInterface, and we can use that to do workgroup level distribution. My plan is using it in the prototype, and we can branch a new op when it becomes tricky.

(cc @lialan @pashu123 )

Reply:
I am strongly -1 to another attribute on the `multi_mma` op. I would just add a new kind of attribute, maybe `#iree_gpu.data_tiled_layout<#iree_gpu.mma_layout<MFMA_F32_16x16x4_F32>>` that captures what you need. Also the permutation fields simply describe the permutation of the input operands relative to the "canonical" layout. The meaning of those permutation fields depend on the `mma_attr` kind on the `multi_mma` op.

Reply:
The permutation is broken if we expand the inner tiles. They are no longer a single M/N/K dimension. The `#iree_gpu.data-tiled_layout<*>` looks better to me, but note that the permutation attribute would be broken. Or we can note that the permutations are not reasonable when the kind is data_tiled_layout. (My concern is that the op semantic is more complicated, but I agree that it is the easiest way to make it work.)

Reply:
The permutation attribute is not restricted to 2-D, in fact it gets expanded to >2D during the course of normal lowering without data tiling already, so expanding the inner tiles during MaterializeEncodings should be fine. AFAICR I didn't write any patterns w/ the multi-mma op that look at the dimensionality or shape of the inner dims; all of that logic is part of interface implementations of `#iree_gpu.mma_layout`. hence if we need different behavior for data tiling, then a new kind of mma_attr like `iree_gpu.data_tiled_layout<...>` is the way to go. Then the permutation fields are only there if they are useful, which perhaps they won't be for data tiling.

Reply:
okay, cool! I'll try to add the attribute and expand the shapes during materialization.

Reply:
https://github.com/iree-org/iree/pull/18294 is ready for review. I'll rebase the PR.